infra/src/main/kotlin/fi/fta/geoviite/infra/geocoding/GeocodingDao.kt

```diff
@@ -55,121 +43,87 @@ class GeocodingDao(
+        val official = getGeocodingContextCacheKey(OFFICIAL, trackNumberId)
+        val trackNumberVersion = versions.findTrackNumber(trackNumberId)?.draftVersion ?: official?.trackNumberVersion
+        // We have to fetch the actual objects (reference line & km-post) here to check references
+        // However, when this is done, the objects are needed elsewhere as well -> they should always be in cache
+        val referenceLineVersion = versions.referenceLines
+            .find { v -> referenceLineDao.fetch(v.draftVersion).trackNumberId == trackNumberId }?.draftVersion
+            ?: official?.referenceLineVersion
+        return if (trackNumberVersion != null && referenceLineVersion != null) {
+            val officialKmPosts = official?.kmPostVersions?.filter { v -> !versions.containsKmPost(v.id) } ?: listOf()
+            val draftKmPosts = versions.kmPosts.filter { draftPost ->
+                kmPostDao.fetch(draftPost.draftVersion).trackNumberId == trackNumberId
+            }.map { v -> v.draftVersion }
+            val kmPostVersions = (officialKmPosts + draftKmPosts).sortedBy { p -> p.id.intValue }
+            GeocodingContextCacheKey(trackNumberVersion, referenceLineVersion, kmPostVersions)
+        } else null
     }
 
     @Cacheable(CACHE_GEOCODING_CONTEXTS, sync = true)
     fun getGeocodingContext(key: GeocodingContextCacheKey): GeocodingContext? {
         logger.daoAccess(AccessType.FETCH, GeocodingContext::class, "cacheKey" to key)
-        return referenceLineDao.fetchVersion(key.publishType, key.trackNumberId)?.let { referenceLineVersion ->
-            val trackNumber = trackNumberDao.fetch(trackNumberDao.fetchVersionOrThrow(key.trackNumberId, key.publishType))
-            val referenceLine = referenceLineDao.fetch(referenceLineVersion)
-            val alignment = referenceLine.alignmentVersion?.let(alignmentDao::fetch)
-                ?: throw IllegalStateException("DB ReferenceLine should have an alignment")
-            val kmPosts = key.fetchVersions(kmPostDao).map(kmPostDao::fetch)
-            if (alignment.segments.isEmpty()) { // If reference line has no geometry, we cannot geocode.
-                null
-            } else {
-                GeocodingContext.create(trackNumber, referenceLine, alignment, kmPosts)
-            }
-        }
+        val trackNumber = trackNumberDao.fetch(key.trackNumberVersion)
+        val referenceLine = referenceLineDao.fetch(key.referenceLineVersion)
+        val alignment = referenceLine.alignmentVersion?.let(alignmentDao::fetch)
+            ?: throw IllegalStateException("DB ReferenceLine should have an alignment")
+        // If the tracknumber is deleted or reference line has no geometry, we cannot geocode.
+        if (!trackNumber.exists || alignment.segments.isEmpty()) return null
+        val kmPosts = key.kmPostVersions.map(kmPostDao::fetch)
+            .sortedBy { post -> post.kmNumber }
+            .filter { post -> post.exists }
```

Comment:
SQL-haussa deletoidut suodatetaan suoraan pois, mutta julkaistujen mukaanotto tekee homman vain versioilla eikä tiedä mikä on poistunut -> lopputuloksessa voi olla poistettuja km-pylväitä